File: algo_py/boj/bj2188.py
import sys
input = sys.stdin.readline


def main():
    def dfs(i):
        for want in cow[i]:
            if check[want]:
                continue
            check[want] = True
            if house[want] == 0 or dfs(house[want]):
                house[want] = i
                return True
        return False

    n, m = map(int, input().split())
    house = [0]*(m+1)
    res = 0
    cow = {i: list(map(int, input().split()))[1:] for i in range(1, n+1)}

    for i in range(1, n+1):
        check = [False]*(m+1)
        if dfs(i):
            res += 1
    print(res)


if __name__ == '__main__':
    main()

File: algo_cpp/bj2188.cpp
#include<cstdio>
#include<vector>
using namespace std;

int house[201];
bool check[201];
vector<int> cow[201];

bool dfs(int i) {
	for (auto w : cow[i]) {
		if (check[w])
			continue;
		check[w] = 1;
		if (house[w] == 0 || dfs(house[w])) {
			house[w] = i;
			return true;
		}
	}
	return false;
}

int main(void)
{
	int n, m, res = 0;
	scanf("%d%d", &n, &m);
	for (int i = 1; i <= n; i++) {
		int s, tmp;
		scanf("%d", &s);
        cow[i].resize(s);
		for (int j = 0; j < s; j++) {
			scanf("%d", &cow[i][j]);
		}
	}
	for (int i = 1; i <= n; i++) {
		for (int i = 1; i <= m; i++)
			check[i] = 0;
		if (dfs(i))
			res += 1;
	}
	printf("%d", res);
	return 0;
}